drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * look for a suitable primary plane for it. After that, we retrieve connector,
+ * CRTC and plane objects properties from the DRM driver. These objects are used
+ * during the atomic modeset setup (see modeset_atomic_modeset()) and also
+ * during the page-flips (see modeset_draw_dev() and modeset_atomic_commit()).
+ */
+
+static int modeset_setup_dev(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev)
+{
+	int ret;
+
+	/* check if a monitor is connected */
+	if (conn->connection != DRM_MODE_CONNECTED) {
+		fprintf(stderr, "ignoring unused connector %u\n",
+			conn->connector_id);
+		return -ENOENT;
+	}
+
+	/* check if there is at least one valid mode */
+	if (conn->count_modes == 0) {
+		fprintf(stderr, "no valid mode for connector %u\n",
+			conn->connector_id);
+		return -EFAULT;
+	}
+
+	/* copy the mode information into our device structure and into
+	 * both buffers */
+	memcpy(&dev->mode, &conn->modes[0], sizeof(dev->mode));
+	dev->bufs[0].width = conn->modes[0].hdisplay;
+	dev->bufs[0].height = conn->modes[0].vdisplay;
+	dev->bufs[1].width = conn->modes[0].hdisplay;
+	dev->bufs[1].height = conn->modes[0].vdisplay;
+	fprintf(stderr, "mode for connector %u is %ux%u\n",
+		conn->connector_id, dev->bufs[0].width, dev->bufs[0].height);
+
+	/* find a crtc for this connector */
+	ret = modeset_find_crtc(fd, res, conn, dev);
+	if (ret) {
+		fprintf(stderr, "no valid crtc for connector %u\n",
+			conn->connector_id);
+		return ret;
+	}
+
+	/* with a connector and crtc, find a primary plane */
+	ret = modeset_find_plane(fd, dev);
+	if (ret) {
+		fprintf(stderr, "no valid plane for crtc %u\n", dev->crtc_id);
+		return ret;
+	}
+
+	/* gather properties of our connector, CRTC and planes */
+	ret = modeset_setup_objects(fd, dev);
+	if (ret) {
+		fprintf(stderr, "cannot get plane properties\n");
+		return ret;
+	}
+
+	/* create framebuffer #1 for this CRTC */
+	ret = modeset_create_fb(fd, &dev->bufs[0]);
+	if (ret) {
+		fprintf(stderr, "cannot create framebuffer for connector %u\n",
+			conn->connector_id);
+		return ret;
+	}
+
+	/* create framebuffer #2 for this CRTC */
+	ret = modeset_create_fb(fd, &dev->bufs[1]);
+	if (ret) {
+		fprintf(stderr, "cannot create framebuffer for connector %u\n",
+			conn->connector_id);
+		modeset_destroy_fb(fd, &dev->bufs[0]);
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * modeset_find_crtc() stays the same.
+ */
+
+static int modeset_find_crtc(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev)
+{
+	drmModeEncoder *enc;
+	unsigned int i, j;
+	uint32_t crtc;
+	struct modeset_dev *iter;
+
+	/* first try the currently conected encoder+crtc */
+	if (conn->encoder_id)
+		enc = drmModeGetEncoder(fd, conn->encoder_id);
+	else
+		enc = NULL;
+
+	if (enc) {
+		if (enc->crtc_id) {
+			crtc = enc->crtc_id;
+			for (iter = modeset_list; iter; iter = iter->next) {
+				if (iter->crtc_id == crtc) {
+					crtc = -1;
```

Comment:
`crtc` is unsigned, so this actually assigns a large positive value here if anything.

Reply:
Make sure to fix *all* code after you changed the type from signed to unsigned. I don't tend point out every single instance of the same mistake separately.

Reply:
Oops, sorry for that.